### PR TITLE
[JVM IR] Maintain KT-36625 bug compatibility between non-IR and IR backends

### PR DESCRIPTION
**Pull request by punzki**
...by removing `IMPLICIT_NOTNULL` casts from `IrStringConcatenation` arguments.

Also fixed an issue where `IrStringConcatenation` can be lowered into a `null` String instead of a literal `"null"` String if the lone argument was a platform type String or String with enhanced nullability and the value was null (e.g., `"${FromJava.nullPlatformString()}"`).

### CODE REVIEW

File: compiler/testData/codegen/box/strings/stringFromJavaPlus.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    }
+}
+
+// FILE: main.kt
+import kotlin.test.assertEquals
+
+fun box(): String {
+    val n = 123
+
+    // Due to KT-36625, certain concatenation calls below (marked with a comment) SHOULD fail a nullability check but do not.
```
Comment (dnpetrov) [inline compiler/testData/codegen/box/strings/stringFromJavaPlus.kt]: Minor: not clear if it's really "SHOULD". Language design formalities are not settled in this case.
I'd change this comment to something like 
```
// Null check behavior in string concatenation might change depending on language design decision for KT-36625.
```

File: compiler/testData/codegen/box/strings/stringFromJavaPlus.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        return "foo";
+    }
+
+    @Nullable
+    public static String nullableStringIsNotNull() {
+        return "foo";
+    }
+}
+
+// FILE: main.kt
```
Comment (dnpetrov) [inline compiler/testData/codegen/box/strings/stringFromJavaPlus.kt]: Minor: I usually give the "main" file in the multi-file test the same name as the test file (`stringFromJavaPlus.kt` in this case). Enables navigation from stack trace in IDEA.

**Comment by dnpetrov** [comment]
Looks ok to me. I'll run tests and push manually.

**Comment by dnpetrov** [comment]
Pushed manually ( ba606147c95edb9e3988654fb44bbe1ac4b24bfe ).

**Comment by punzki** [comment]
@dnpetrov Thanks! I don't know if you saw this but added an extra commit on the branch to clarify the comments about KT-36625. Feel free to merge, ignore, or replace with your own.

**Comment by dnpetrov** [comment]
Thank you! Yes, I see, I've missed it. I'll push it as a separate commit.

**Comment by dnpetrov** [comment]
Done ( 89cf32eccc813b10b506799ae132f1fb5dbf03a9 )

**Pull request closed without merging**